lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2609 @@
+{
+    struct txn *this = *ptr;
+    struct txn *next = this->next;
+    this->file->refcount--;
+    free(this);
+    *ptr = next;
+}
+
+static void _remove_file(struct tm_file **ptr)
+{
+    struct tm_file *this = *ptr;
+    struct tm_file *next = (*ptr)->next;
+    assert(!this->refcount);
+    assert(!this->has_lock);
+    if (this->base) munmap(this->base, this->size);
+    if (this->fd != -1) close(this->fd);
+    free(this);
+    *ptr = next;
+}
+
+static void empty_db(struct dbengine *db)
+{
+    if (db->loc.file) {
+        db->loc.file->refcount--;
+        db->loc.file = NULL;
+    }
+    while (db->write_txn)
+        _remove_txn(&db->write_txn);
+    while (db->read_txn)
+        _remove_txn(&db->read_txn);
+    while (db->openfile)
+        _remove_file(&db->openfile);
+}
+
+static void tm_cleanup(struct dbengine *db)
+{
+    // close any open files with a zero refcount (no loc or txn
+    // still points there) except the first one
+    if (!db->openfile) return;
+    struct tm_file **ptr = &db->openfile->next;
+    while (*ptr) {
+        if (!(*ptr)->refcount) {
+            _remove_file(ptr);
+        }
+        else {
+            ptr = &((*ptr)->next);
+        }
+    }
+}
+
+/* end cleanup */
+
+static inline int tm_commit(struct dbengine *db, size_t len)
+{
+    assert(db->openfile);
+    if (!db->openfile->dirty) return 0;
+    assert(db->openfile->has_lock == 2);
+    db->openfile->dirty = 0;
+    if (db->nosync) return 0;
+    return msync(db->openfile->base, len, MS_SYNC|MS_INVALIDATE);
+}
+
+static inline int tm_ensure(struct dbengine *db, size_t offset)
+{
+    struct tm_file *file = db->openfile;
+    if (offset <= file->size) return 0;
+
+    assert(file);
+    assert(file->has_lock == 2);
+
+    offset = tm_roundup(offset);
+
+    // make sure anything we have mapped in has committed to disk
+    if (file->dirty) {
+        if (!db->nosync && msync(file->base, file->size, MS_ASYNC)) {
+            xsyslog(LOG_ERR, "DBERROR: twom failed to msync during tm_ensure",
+                             "filename=<%s> size=<%08llX>",
+                             db->fname, (LLU)file->size);
+            return CYRUSDB_IOERROR;
+        }
+        file->dirty = 0;
+    }
+
+    if (ftruncate(file->fd, offset)) {
+        xsyslog(LOG_ERR, "DBERROR: twom failed to extend file during tm_ensure",
+                         "filename=<%s> size=<%08llX> newsize=<%08llX>",
+                         db->fname, (LLU)file->size, (LLU)offset);
+        return CYRUSDB_IOERROR;
+    }
+
+    // map the larger file into new memory
+    if (file->size) munmap(file->base, file->size);
+    file->base = mmap((caddr_t)0, offset, PROT_READ|PROT_WRITE, MAP_SHARED, db->openfile->fd, 0L);
+    file->size = offset;
+
+    return 0;
+}
+
+/* given an open, mapped, locked db, write the header information */
+static inline int write_header(struct dbengine *db, struct tm_header *header)
+{
+    int r = tm_ensure(db, HEADER_SIZE);
+    if (r) return r;
```

Comment:
The current and the next function `commit_header` can easily be rewritten not to use variable `r`.

Reply:
I don't care about that.  Easier to debug with variables and the compiler will clean them up.  Readability trumps micro optimisation in this case, it's only called twice per commit.  I'm all for micro-optimising advance_loc and other inner loop stuff, but not this.